File: src/main/java/com/lcy/java/spark/rdd/instance/GroupTopN.java
package com.lcy.java.spark.rdd.instance;

import org.apache.commons.collections.IteratorUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

//此时取出来的top3 不是排序后的前三个，而是随机的三个

public class GroupTopN {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("study").setMaster("local");
        JavaSparkContext jsc = new JavaSparkContext(conf);

        jsc.textFile("file:\\D:\\sparkData\\demo.txt")
                .mapToPair(line ->{
                    String[] strings = line.split(",");
                    return new Tuple2<>(strings[0], Integer.parseInt(strings[1]));
                })
                .groupByKey()
                .mapToPair(e -> new Tuple2<>(e._1, IteratorUtils.toList(e._2.iterator()).subList(0, 3)))
                .foreach(e -> System.out.println(e._1 + " " + e._2));
        jsc.stop();

    }

}

/**
 * 数据
 * class1 90
 * class2 56
 * class1 87
 * class1 76
 * class2 88
 * class1 95
 * class1 74
 * class2 87
 * class2 67
 * class2 77
 */

File: src/main/scala/com/lcy/scala/spark/rdd/instance/GroupTopN.scala
package com.lcy.scala.spark.rdd.instance

import org.apache.spark.{SparkConf, SparkContext}

object GroupTopN {

    def main(args: Array[String]): Unit = {

        val conf = new SparkConf().setAppName("study").setMaster("local")
        val sc = new SparkContext(conf)

        sc.textFile("file:\\D:\\sparkData\\demo.txt")
                .map(line => {
                    val strings = line.split(",")
                    (strings(0), strings(1).toInt)
                })
                .sortBy(e => e._2, false)
                .groupByKey() //排序倒序
                .map(e => (e._1, e._2.take(3)))
                .foreach(e => {
                    print(e._1 + ":")
                    e._2.foreach(e => print(e + " "))
                    println()
                })
        sc.stop()
    }

}


/**
 * 数据
 *
 * class2,1
 * class3,2
 * class3,11
 * class4,12
 * class2,13
 * class3,14
 * class4,15
 * class3,5
 * class4,6
 * class2,7
 * class4,3
 * class2,4
 * class3,8
 * class4,9
 * class2,10
 *
 */
